### ISSUE TITLE
test model

### ISSUE THREAD

**Issue by territoryLiu** [ShaohonChen/Qwen3-SmVL#3]
I wrote a test script, but it failed to run. Can it fade out of the test after training
`import os
from dataclasses import dataclass, field
from PIL import Image
import torch
from transformers import (
    AutoProcessor,
    AutoModelForImageTextToText,
    AutoModelForCausalLM,
    AutoTokenizer
)
from functools import partial

# ========= connector & config 定义 =========
class SmolVLMConnector(torch.nn.Module):
    def __init__(self, config):
        super().__init__()
        self.config = config
        self.linear = torch.nn.Linear(
            config.vision_config.hidden_size,
            config.text_config.hidden_size
        )

    def forward(self, x):
        return self.linear(x)

@dataclass
class VisionConfig:
    hidden_size: int = 768

@dataclass
class TextConfig:
    hidden_size: int = 1024

@dataclass
class ConnectConfig:
    scale_factor: int = 4
    vision_config: VisionConfig = field(default_factory=VisionConfig)
    text_config: TextConfig = field(default_factory=TextConfig)

# ========= 加载模型 =========
def load_fused_model(model_dir, device="cuda:0"):
    # 加载主模型
    model = AutoModelForImageTextToText.from_pretrained(
        "model/SmolVLM2-256M-Video-Instruct",
        torch_dtype=torch.bfloat16,
        _attn_implementation="eager",
    ).to(device)

    # 加载 Qwen LLM 模型（注意：要与训练时一致）
    qwen = AutoModelForCausalLM.from_pretrained(
        "model/Qwen3-0.6B",  # 必须跟训练时加载路径一样
        torch_dtype=torch.bfloat16,
    ).to(device)

    # 替换 text model & lm head
    model.model.text_model = qwen.model
    model.lm_head = qwen.lm_head
    model.vocab_size = qwen.vocab_size
    model.model.vocab_size = qwen.vocab_size
    model.config.vocab_size = qwen.vocab_size
    model.config.text_config.vocab_size = qwen.vocab_size
    model.model.config.vocab_size = qwen.vocab_size
    model.model.config.text_config.vocab_size = qwen.vocab_size

    # 替换 image_token_id 和 eos_token_id（同训练）
    model.image_token_id = 151655
    model.model.image_token_id = 151655
    model.config.image_token_id = 151655
    model.model.config.image_token_id = 151655
    model.generation_config.eos_token_id = 151645

    # 创建并挂载 connector
    connector_cfg = ConnectConfig()
    connector = SmolVLMConnector(connector_cfg).to(device).to(torch.bfloat16)
    model.model.connector = connector

    return model

# ========= 加载 Processor =========
def load_processor(model_dir):
    # return AutoProcessor.from_pretrained("model/SmolVLM2-256M-Video-Instruct")
    smolvlm2_processor = AutoProcessor.from_pretrained(
        "model/SmolVLM2-256M-Video-Instruct"
    )
    qwen3_tokenizer = AutoTokenizer.from_pretrained("model/Qwen3-0.6B")

    smolvlm2_processor.tokenizer = qwen3_tokenizer
    with open("chat_template.jinja", "r") as f:
        smolvlm2_processor.chat_template = f.read()
    smolvlm2_processor.fake_image_token = "<vision_start>"
    smolvlm2_processor.image_token = "<|image_pad|>"
    smolvlm2_processor.image_token_id = 151655
    smolvlm2_processor.end_of_utterance_token = "<im_end>"
    smolvlm2_processor.global_image_token = "<|vision_pad|>"
    smolvlm2_processor.video_token = "<|video_pad|>"

    return smolvlm2_processor

# ========= 推理函数 =========
def inference(model_dir, image_path, question, device="cuda:0"):
    model = load_fused_model(model_dir, device=device)
    processor = load_processor(model_dir)

    messages = [
        {"role": "system", "content": "使用中文回答所有问题。"},
        {
            "role": "user",
            "content": [
                {"type": "image"},
                {"type": "text", "text": question},
            ],
        },
    ]

    # 构建输入
    input_text = processor.apply_chat_template(
        messages,
        add_generation_prompt=True,
        tokenize=False,
        enable_thinking=True,
    )

    print("############ 输入文本 ############")
    print(input_text)

    # 加载图像
    image = Image.open(image_path).convert("RGB")
    # 调整图像大小，确保宽高都是16的倍数
    width, height = image.size
    new_width = (width // 16) * 16
    new_height = (height // 16) * 16
    if width != new_width or height != new_height:
        image = image.resize((new_width, new_height))
    images = [[image]]  # 注意要套两层 list，和训练时保持一致

    # 编码
    batch = processor(
        text=[input_text],
        images=images,
        max_length=1024,
        return_tensors="pt",
        padding_side="left",
        padding=True,
    ).to(device, dtype=torch.bfloat16)

    # 生成
    with torch.no_grad():
        generated_ids = model.generate(
            **batch,
            do_sample=False,
            max_new_tokens=256,
        )

    input_len = batch["input_ids"].shape[1]

    # 解码
    full_output = processor.batch_decode(generated_ids, skip_special_tokens=False)
    generated_texts = processor.batch_decode(
        generated_ids[:, input_len:], skip_special_tokens=True
    )

    print("############ 模型输出 ############")
    print(generated_texts[0])

# ========= 主程序 =========
if __name__ == "__main__":
    model_dir = "model/freeze_llm_vlm_cocoqa"   # 替换为你保存模型的目录
    image_path = "panda.png"              # 替换为你要推理的图像路径
    question = "图中有什么动物？"       # 替换为你想问的问题

    inference(model_dir, image_path, question)
`

<img width="849" height="215" alt="Image" src="https://github.com/user-attachments/assets/8ef20b0b-cdfd-4d63-af2e-da9c72a7be04" />

**Comment by ShaohonChen** [comment]
能否重新用"```"包装一下代码，不然我没看懂你的代码😂，非常欢迎提一个贡献，我可以帮你一起看
